File: crates/oharness-loop/examples/custom_critic.rs
//! `custom_critic` — implement the `Critic` trait from scratch and
//! attach it to an agent.
//!
//! Shows:
//! 1. What implementing `Critic` looks like end-to-end (it's tiny —
//!    one trait method, one verdict value).
//! 2. How a `Reject` verdict surfaces at the loop layer: the run
//!    terminates with `Termination::Failed { category: Critic }` and
//!    a `critic.rejected` event lands in the trajectory.
//!
//! The shipped critics (`LlmJudgeCritic`, `TestCritic`,
//! `RegexDenyCritic`, `ConstitutionalCritic`) all follow the same
//! pattern; this example is the template for writing your own.
//!
//! Run with:
//!
//! ```bash
//! cargo run --example custom_critic -p oharness-loop
//! ```

use async_trait::async_trait;
use oharness_core::event::EventKind;
use oharness_core::{
    CompletionRequest, CompletionResponse, Content, LlmCapabilities, ModelId, StopReason, Task,
    Termination, Usage,
};
use oharness_critic::{
    AggregationPolicy, AssessmentContext, CompositeCritic, Critic, CriticVerdict,
};
use oharness_llm::{ChunkStream, Llm, LlmError};
use oharness_loop::{Agent, ReactLoop};
use oharness_tools::fs::FsToolSet;
use oharness_trace::InMemorySink;
use std::sync::Arc;

// ---------------------------------------------------------------------
// Custom critic: rejects any assistant turn that hedges.
//
// Real critics care about more than one phrase — the pattern is the
// same: look at `ctx.latest_turn.message`, inspect the text blocks,
// emit a verdict.
// ---------------------------------------------------------------------

struct NoHedgingCritic;

#[async_trait]
impl Critic for NoHedgingCritic {
    fn name(&self) -> &str {
        "no-hedging"
    }

    async fn assess(&self, ctx: &AssessmentContext<'_>) -> CriticVerdict {
        // The `AssistantTurn` carries `message`, which is a
        // `Message::Assistant { content, .. }`. We scan the text
        // blocks for hedge-words.
        let oharness_core::Message::Assistant { content, .. } = &ctx.latest_turn.message else {
            // Not an assistant turn — nothing to assess. Accept.
            return CriticVerdict::Accept;
        };

        let joined_text = content
            .iter()
            .filter_map(|c| match c {
                Content::Text { text } => Some(text.as_str()),
                _ => None,
            })
            .collect::<Vec<_>>()
            .join(" ")
            .to_ascii_lowercase();

        const HEDGES: &[&str] = &["i'm not sure", "i am not sure", "maybe", "possibly"];
        for hedge in HEDGES {
            if joined_text.contains(hedge) {
                return CriticVerdict::Reject {
                    reason: format!("response hedges: found '{hedge}'"),
                };
            }
        }
        CriticVerdict::Accept
    }
}

// ---------------------------------------------------------------------
// Scripted LLM that hedges on its one and only turn — so the critic
// fires. Swap in a confident response to see the `Accept` path.
// ---------------------------------------------------------------------

struct ScriptedHedgeLlm;

#[async_trait]
impl Llm for ScriptedHedgeLlm {
    fn name(&self) -> &str {
        "scripted"
    }

    fn capabilities(&self) -> LlmCapabilities {
        LlmCapabilities::default()
    }

    async fn complete(&self, _req: CompletionRequest) -> Result<CompletionResponse, LlmError> {
        Ok(CompletionResponse {
            id: "msg_1".into(),
            model: ModelId::new("scripted-hedger"),
            content: vec![Content::text("I'm not sure what you're asking.")],
            stop_reason: StopReason::EndTurn,
            usage: Usage {
                tokens_input: 8,
                tokens_output: 9,
                ..Default::default()
            },
        })
    }

    async fn stream(&self, _req: CompletionRequest) -> Result<ChunkStream, LlmError> {
        Err(LlmError::Unsupported("stream"))
    }
}

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    // `CompositeCritic` is the entry point — a single critic still
    // goes through it so the aggregation policy is explicit. Use
    // `AggregationPolicy::FirstReject` for fail-fast behavior, or
    // `AllMustAccept` / `MajorityVote` for voting semantics.
    let critics = Arc::new(
        CompositeCritic::new("hedge-guard", AggregationPolicy::FirstReject)
            .push(Box::new(NoHedgingCritic)),
    );

    let sink = Arc::new(InMemorySink::new());
    let agent = Agent::builder()
        .with_llm(Arc::new(ScriptedHedgeLlm))
        .with_tools(Arc::new(FsToolSet::new()))
        .with_event_sink(sink.clone())
        .with_loop(Box::new(ReactLoop::new()))
        .with_critics(critics)
        .with_max_turns(3)
        .build()?;

    let outcome = agent.run(Task::new("figure it out")).await?;

    // The critic rejects on turn 1 → run fails with category=Critic.
    match &outcome.termination {
        Termination::Failed { error, .. } => {
            println!("Termination: Failed (category={:?})", error.category);
            println!("Critic message: {}", error.message);
        }
        other => println!("Termination: {other:?}"),
    }

    // The trajectory carries a `critic.rejected` event so downstream
    // analysis (or Reflexion) can see exactly what the critic saw.
    let rejections = sink
        .events()
        .iter()
        .filter(|e| matches!(e.kind, EventKind::CriticRejected(_)))
        .count();
    println!("critic.rejected events: {rejections}");
    Ok(())
}

File: crates/oharness-py/examples/custom_critic.py
"""custom_critic — implement the `Critic` trait from scratch.

Shows:
1. What implementing `Critic` looks like end-to-end (one method,
   one verdict JSON string).
2. How a `reject` verdict surfaces at the loop layer: the run
   terminates with `Termination::Failed { category: Critic }`.

Python can emit four verdicts: `accept`, `accept_with_note`,
`reject`, `abort`. The `revise` variant (in-place turn rewrite)
is not currently supported from Python — see `self_refine.py`
for the trade-off, or `oharness-loop/examples/self_refine.rs`
for the Rust version.
"""

import json

import oharness

HEDGES = ("i'm not sure", "i am not sure", "maybe", "possibly")


class NoHedgingCritic:
    """Reject any assistant turn containing hedge phrases.

    The Python class gets an `assess(ctx_json) -> str` method;
    `ctx_json` is a trimmed `AssessmentContext` carrying
    `task`, `latest_turn` (the just-emitted assistant message),
    and `turn_index`. Returns a verdict JSON.
    """

    def assess(self, ctx_json: str) -> str:
        ctx = json.loads(ctx_json)
        turn = ctx.get("latest_turn") or {}
        content = turn.get("content", [])
        text_blocks = [
            b.get("text", "") for b in content
            if b.get("type") == "text"
        ]
        joined = " ".join(text_blocks).lower()
        for hedge in HEDGES:
            if hedge in joined:
                return json.dumps({
                    "verdict": "reject",
                    "reason": f"response hedges: found '{hedge}'",
                })
        return json.dumps({"verdict": "accept"})


class ScriptedHedgeLlm:
    """Scripted LLM that hedges on its one and only turn."""

    def complete(self, req_json: str) -> str:
        return json.dumps({
            "id": "msg_1",
            "model": "scripted-hedger",
            "content": [{"type": "text", "text": "I'm not sure what you're asking."}],
            "stop_reason": {"kind": "end_turn"},
            "usage": {"tokens_input": 8, "tokens_output": 9},
        })


def main() -> None:
    # A CompositeCritic wraps one or more critics with an
    # aggregation policy. Even a single critic goes through the
    # composite so the policy is explicit.
    critics = oharness.CompositeCritic("hedge-guard", "first_reject")
    critics.push(oharness.PyCritic(NoHedgingCritic(), name="no-hedging"))

    sink = oharness.InMemorySink()
    agent = (
        oharness.Agent.builder()
        .with_llm(oharness.PyLlm(ScriptedHedgeLlm(), name="scripted"))
        .with_tools(oharness.FsToolSet())
        .with_event_sink(sink)
        .with_loop(oharness.ReactLoop())
        .with_critics(critics)
        .with_max_turns(3)
        .build()
    )

    outcome = json.loads(agent.run(oharness.Task("figure it out")))

    # Reject on turn 1 → run fails with category=Critic.
    termination = outcome["termination"]
    print(f"Termination: {termination}")
    if termination.get("kind") == "failed":
        err = termination.get("error", {})
        print(f"Critic message: {err.get('message')}")
        print(f"Category: {err.get('category')}")

    # The trajectory carries a `critic.rejected` event. Events
    # are shaped as `{v, seq, run_id, timestamp, span_id, type,
    # payload, ...}` — `type` is the event kind tag.
    events = json.loads(sink.events_json())
    rejections = sum(1 for e in events if e.get("type") == "critic.rejected")
    print(f"critic.rejected events: {rejections}")


if __name__ == "__main__":
    main()
